Mathlib/Probability/Independence/Basic.lean

```diff
@@ -0,0 +1,923 @@
+    have h_t1 : t1 = ⋂ n ∈ s, f n := by
+      suffices h_forall : ∀ n ∈ s, f n = ft1 n
+      · rw [ht1_eq]
+        ext x
+        simp_rw [Set.mem_iInter]
+        conv => lhs; intro i hns; rw [← h_forall i hns]
+      intro n hnS
+      have hn_ne_a : n ≠ a := by rintro rfl; exact haS (hs_mem hnS)
+      simp_rw [if_pos hnS, if_neg hn_ne_a]
+    have h_μ_t1 : μ t1 = ∏ n in s, μ (f n) := by rw [h_t1, ← hp_ind s h_f_mem_pi]
+    have h_t2 : t2 = f a := by simp
+    have h_μ_inter : μ (t1 ∩ t2) = ∏ n in insert a s, μ (f n) := by
+      have h_t1_inter_t2 : t1 ∩ t2 = ⋂ n ∈ insert a s, f n := by
+        rw [h_t1, h_t2, Finset.set_biInter_insert, Set.inter_comm]
+      rw [h_t1_inter_t2, ← hp_ind (insert a s) h_f_mem]
+    have has : a ∉ s := fun has_mem => haS (hs_mem has_mem)
+    rw [h_μ_inter, Finset.prod_insert has, h_t2, mul_comm, h_μ_t1]
+set_option linter.uppercaseLean3 false in
+#align probability_theory.Indep_sets.pi_Union_Inter_of_not_mem ProbabilityTheory.iIndepSets.piiUnionInter_of_not_mem
+
+/-- The measurable space structures generated by independent pi-systems are independent. -/
+theorem iIndepSets.iIndep [ProbabilityMeasure μ] (m : ι → MeasurableSpace Ω) (h_le : ∀ i, m i ≤ m0)
+    (π : ι → Set (Set Ω)) (h_pi : ∀ n, IsPiSystem (π n))
+    (h_generate : ∀ i, m i = generateFrom (π i)) (h_ind : iIndepSets π μ) : iIndep m μ := by
+  classical
+    intro s f
+    refine Finset.induction ?_ ?_ s
+    · simp only [Finset.not_mem_empty, Set.mem_setOf_eq, IsEmpty.forall_iff, implies_true,
+      Set.iInter_of_empty, Set.iInter_univ, measure_univ, Finset.prod_empty, forall_true_left]
+    · intro a S ha_notin_S h_rec hf_m
+      have hf_m_S : ∀ x ∈ S, MeasurableSet[m x] (f x) := fun x hx => hf_m x (by simp [hx])
+      rw [Finset.set_biInter_insert, Finset.prod_insert ha_notin_S, ← h_rec hf_m_S]
+      let p := piiUnionInter π S
+      set m_p := generateFrom p with hS_eq_generate
+      have h_indep : @Indep Ω m_p (m a) m0 μ := by
+        have hp : IsPiSystem p := isPiSystem_piiUnionInter π h_pi S
+        have h_le' : ∀ i, generateFrom (π i) ≤ m0 := fun i => (h_generate i).symm.trans_le (h_le i)
+        have hm_p : m_p ≤ m0 := generateFrom_piiUnionInter_le π h_le' S
+        exact IndepSets.indep hm_p (h_le a) hp (h_pi a) hS_eq_generate (h_generate a)
+            (iIndepSets.piiUnionInter_of_not_mem h_ind ha_notin_S)
+      refine @Indep.symm Ω _ _ m0 μ h_indep (f a) (⋂ n ∈ S, f n)
+        (hf_m a (Finset.mem_insert_self a S)) ?_
+      have h_le_p : ∀ i ∈ S, m i ≤ m_p := by
+        intro n hn
+        rw [hS_eq_generate, h_generate n]
+        exact le_generateFrom_piiUnionInter _ hn
+      have h_S_f : ∀ i ∈ S, MeasurableSet[m_p] (f i) :=
+        fun i hi => (h_le_p i hi) (f i) (hf_m_S i hi)
+      exact S.measurableSet_biInter h_S_f
+set_option linter.uppercaseLean3 false in
+#align probability_theory.Indep_sets.Indep ProbabilityTheory.IndepSets.indep
```

Comment:
This align is wrong. That just caused me some trouble 😢

Reply:
Oops, sorry about that. To be honest, I don't even remember writing this PR... 😞